charms/reconciler.py

```diff
@@ -1,26 +1,36 @@
 import logging
+from typing import Callable, Optional
 
-from ops import HookEvent, Object, StoredState, UpdateStatusEvent
+from ops import BlockedStatus, BoundEvent, HookEvent, Object, CharmBase, EventBase, StatusBase
 
-import charms.contextual_status as status
 
 log = logging.getLogger(__name__)
 
 
-class Reconciler(Object):
-    stored = StoredState()
+class ReconcileErrorWithStatus(Exception):
+    def __init__(
+            self: "ReconcileErrorWithStatus", 
+            message: str, 
+            status: StatusBase,
+        ) -> None:
+        self.status = status
+        self.message = message
+        super().__init__(message)
+
+    def __str__(self: "ReconcileErrorWithStatus") -> str:
+        return f"[{self.status}] {self.message}"
 
+
+class Reconciler(Object):
     def __init__(
-        self,
-        charm,
-        reconcile_function,
-        exit_status=None,
-        custom_events=None,
+        self: "Reconciler",
+        charm: CharmBase,
+        reconcile_function: Callable[[EventBase], None],
+        custom_events: Optional[list[BoundEvent]] = None,
+        exit_status: StatusBase = BlockedStatus("Reconcile failed"),
```

Comment:
Um, let's not swap the order here...  also use `List` b/c py38 is special
```suggestion
        exit_status: StatusBase = BlockedStatus("Reconcile failed"),
        custom_events: Optional[List[BoundEvent]] = None,
```